// @airmcp-dev/core/dist/plugin/builtin/ratelimit-per-user.js
// @airmcp-dev/core — plugin/builtin/ratelimit-per-user.ts
//
// 사용자별 레이트 리밋 플러그인.
// shield의 글로벌 레이트 리밋과 달리, 사용자/세션별 개별 추적.
//
// @example
//   defineServer({
//     use: [perUserRateLimitPlugin({
//       maxCalls: 10,
//       windowMs: 60_000,
//       identifyBy: '_userId',
//     })],
//   });
export function perUserRateLimitPlugin(options) {
    const maxCalls = options?.maxCalls ?? 10;
    const windowMs = options?.windowMs ?? 60_000;
    const identifyBy = options?.identifyBy ?? '_userId';
    const windows = new Map();
    const middleware = {
        name: 'air:per-user-ratelimit',
        before: async (ctx) => {
            const userId = ctx.params[identifyBy] || 'anonymous';
            const now = Date.now();
            const key = `${userId}:${ctx.tool.name}`;
            const timestamps = (windows.get(key) || []).filter((t) => t > now - windowMs);
            if (timestamps.length >= maxCalls) {
                return {
                    abort: true,
                    abortResponse: {
                        content: [{ type: 'text', text: `[RateLimit] User "${userId}" exceeded ${maxCalls} calls per ${windowMs / 1000}s for "${ctx.tool.name}".` }],
                        isError: true,
                    },
                };
            }
            timestamps.push(now);
            windows.set(key, timestamps);
        },
    };
    return {
        meta: { name: 'air:per-user-ratelimit', version: '1.0.0' },
        middleware: [middleware],
    };
}
