// @airmcp-dev/core/dist/plugin/plugin-hook.js
// @airmcp-dev/core — plugin/plugin-hook.ts
// 플러그인 훅 실행기
export class PluginHookRunner {
    hooks = [];
    register(pluginName, hooks) {
        this.hooks.push({ name: pluginName, hooks });
    }
    async runOnInit(ctx) {
        for (const { hooks } of this.hooks) {
            if (hooks.onInit)
                await hooks.onInit(ctx);
        }
    }
    async runOnStart(ctx) {
        for (const { hooks } of this.hooks) {
            if (hooks.onStart)
                await hooks.onStart(ctx);
        }
    }
    async runOnStop(ctx) {
        for (const { hooks } of this.hooks) {
            if (hooks.onStop)
                await hooks.onStop(ctx);
        }
    }
    runOnToolRegister(tool, ctx) {
        let result = tool;
        for (const { hooks } of this.hooks) {
            if (hooks.onToolRegister) {
                const modified = hooks.onToolRegister(result, ctx);
                if (modified)
                    result = modified;
            }
        }
        return result;
    }
}
