// @airmcp-dev/core/dist/storage/memory-store.js
// @airmcp-dev/core — storage/memory-store.ts
// 인메모리 StorageAdapter 구현 (개발용 기본값)
export class MemoryStore {
    data = new Map();
    logs = new Map();
    async init() {
        /* no-op */
    }
    async set(namespace, key, value, ttl) {
        const fullKey = `${namespace}:${key}`;
        const expireAt = ttl ? Date.now() + ttl * 1000 : undefined;
        this.data.set(fullKey, { value, expireAt });
    }
    async get(namespace, key) {
        const fullKey = `${namespace}:${key}`;
        const entry = this.data.get(fullKey);
        if (!entry)
            return null;
        if (entry.expireAt && Date.now() > entry.expireAt) {
            this.data.delete(fullKey);
            return null;
        }
        return entry.value;
    }
    async delete(namespace, key) {
        return this.data.delete(`${namespace}:${key}`);
    }
    async list(namespace, prefix) {
        const nsPrefix = prefix ? `${namespace}:${prefix}` : `${namespace}:`;
        const keys = [];
        for (const k of this.data.keys()) {
            if (k.startsWith(nsPrefix))
                keys.push(k.slice(namespace.length + 1));
        }
        return keys;
    }
    async entries(namespace, prefix) {
        const nsPrefix = prefix ? `${namespace}:${prefix}` : `${namespace}:`;
        const result = [];
        for (const [k, entry] of this.data.entries()) {
            if (!k.startsWith(nsPrefix))
                continue;
            if (entry.expireAt && Date.now() > entry.expireAt)
                continue;
            result.push({ key: k.slice(namespace.length + 1), value: entry.value });
        }
        return result;
    }
    async append(namespace, entry) {
        if (!this.logs.has(namespace))
            this.logs.set(namespace, []);
        this.logs.get(namespace).push({ ...entry, _ts: Date.now() });
    }
    async query(namespace, opts) {
        let entries = this.logs.get(namespace) || [];
        if (opts?.since)
            entries = entries.filter((e) => e._ts >= opts.since.getTime());
        if (opts?.until)
            entries = entries.filter((e) => e._ts <= opts.until.getTime());
        if (opts?.filter) {
            entries = entries.filter((e) => {
                for (const [k, v] of Object.entries(opts.filter)) {
                    if (e[k] !== v)
                        return false;
                }
                return true;
            });
        }
        entries.sort((a, b) => b._ts - a._ts);
        const offset = opts?.offset || 0;
        const limit = opts?.limit || 100;
        return entries.slice(offset, offset + limit);
    }
    async close() {
        this.data.clear();
        this.logs.clear();
    }
}
